lib/saluki-env/src/autodiscovery/providers/local.rs

```diff
@@ -0,0 +1,298 @@
+use std::collections::HashSet;
+use std::path::PathBuf;
+
+use async_trait::async_trait;
+use saluki_error::GenericError;
+use tokio::fs;
+use tokio::sync::broadcast::{self, Receiver, Sender};
+use tokio::sync::OnceCell;
+use tokio::time::{interval, Duration};
+use tracing::{debug, warn};
+
+use crate::autodiscovery::{AutodiscoveryEvent, AutodiscoveryProvider, Config, EventType};
+
+/// A local auto-discovery provider that uses the file system.
+pub struct LocalAutoDiscoveryProvider {
+    search_paths: Vec<PathBuf>,
+    sender: Sender<AutodiscoveryEvent>,
+    listener_init: OnceCell<()>,
+}
+
+impl LocalAutoDiscoveryProvider {
+    /// Creates a new `LocalAutoDiscoveryProvider` that will monitor the specified paths.
+    pub fn new<P: AsRef<std::path::Path>>(paths: Vec<P>) -> Self {
+        let search_paths: Vec<PathBuf> = paths
+            .iter()
+            .filter_map(|p| {
+                if !p.as_ref().exists() {
+                    warn!("Skipping path '{}' as it does not exist", p.as_ref().display());
+                    return None;
+                }
+                if !p.as_ref().is_dir() {
+                    warn!("Skipping path '{}', it is not a directory.", p.as_ref().display());
+                    return None;
+                }
+                Some(p.as_ref().to_path_buf())
+            })
+            .collect();
+
+        let (sender, _) = broadcast::channel::<AutodiscoveryEvent>(100);
```

Comment:
We should lift up this `100` to a constant (for example, `const AD_STREAM_CAPACITY: usize = 100;`) just so that it's a little more discoverable when reading the file.